### ISSUE TITLE
FB: Suggest control name from the control label

### ISSUE THREAD

**Issue by ajw625** [orbeon/orbeon-forms#1960]
It would be nice if entering a label for a control or a section would automatically populate the control name and section name. I never want to use the default name, e.g., "control-47". Automatically setting it from the label if it matches that default pattern would save a lot of work in the long run.

**Comment by ebruchez** [comment]
[+1 from customer](https://mail.google.com/mail/u/0/#inbox/14ff164cb57b506b)

**Comment by ebruchez** [comment]
I like the idea but the devil is in the details.

Just after a control is inserted, it doesn't have a custom label yet. However the control must have a name/id. So at that time, we still need the automatically-generated name.

If the form author then sets a label, we could generate a name automatically based on the label, making names unique with suffixes (`-1`, `-2`, etc.).

Now what if the user sets a custom name, say `first-name`, and then later updates the label? Should the name be changed again?

Should the label change as the user updates the label, or should we do it only the first time the user sets a label?

**Comment by avernet** [comment]
Yes, the devil is in the details. For this to work to work, I think we need the following:

1. To be able to track when form authors "manually" change the control name. I.e. in the Control Settings, we could have an explicit checkbox "custom control name", unchecked by default. When unchecked, the control name shows read-only, and the name is automatically updated when the label is changed. When checked, the control name can be edited, and it isn't changed when the label changes. The status of the checkbox would to be stored somewhere (custom MIP?).
2. When the control name changes, whether automatically (because the label changed) or manually, references to the control done with variables are changed. (But non-variable references, e.g. paths, are not updated.)

**Comment by ebruchez** [comment]
You could also maybe do this: only set the control name based on the label if the current control name looks like an automatically-generated one. So:

- You have `control-42` and user enters `First Name`, then control name is set to `first-name`.
- User changes control name to `name`.
- User changes label to `Last Name`. Control name remains `name`.

This would help a little bit, but might be too limiting and confusing. So I think this would not be as good as what you suggest.

Now if we go your way:

- Where/how would we store the information that the control name is custom?
- How do we handle backward compatibility?

We could maybe just have an option in Form Settings to say that control names are automatically generated. This would be global per form. Existing forms would have the option set  to `false`, new forms to `true`. The user can toggle back and forth as needed.

If we do want to store information about whether the control name should track the label, we have these options:

- attribute on data holder
    - bad solution as we don't want to pollute the data
- attribute on bind
    - this wouldn't be a custom MIP, just an attribute
- attribute or element in resources, maybe `<label update-name="true">`
    - one issue is that this has to be replicated across languages
- attribute on control
   - would need to be in `fb:` namespace
- off-band, maybe in metadata
    - this is not metadata like other metadata, just something for Form Builder

We have issue #765 about updating expressions when control names change, which is orthogonal to the issue of automatically generating control names.

**Comment by avernet** [comment]
I like the ideas of storing this on the `<xf:label>`. And there is only one `<xf:label>` even if the form is in multiple languages. However, if there are multiple languages, which label would you use to update the name with? We could:

1. Allow a different one for each control, allowing users to to select which language to use, with the default being the first language of the form, and store this as `<xf:label update-name-from="en">`.
2. Always use the first language of the form.

**Comment by ebruchez** [comment]
@avernet I was talking about storing this on the `<label>` in the resources, not `<xf:label>` in the control, but maybe your idea is even better.

Good point on multiple languages!

- I don't think allowing to choose per control makes much sense. The elements in the XML data should be in one language, which the form authors should be able to choose.
- Currently we can't reorder languages or choose a main language in FB.
- What if you mess up? You would want a function which says: "update all control names from the English labels now".

**Comment by avernet** [comment]
@ebruchez I agree, it doesn't seem useful to have different controls have their name derive from labels in different languages (e.g. `first-name` and `nom-de-famille`). So what about a single setting in the "Form Settings", allowing form authors to choose the language from which the control names should be derived (for those controls where the control name is marked as "automatic" in the Control Settings, which would be the default, at least for new forms or new controls)?

**Comment by ebruchez** [comment]
@avernet Sure that setting would be ok. At first we probably don't even need it.

**Comment by ebruchez** [comment]
+1 from me today.

**Comment by ebruchez** [comment]
A simpler way to do all this would be to be explicit about it in Control Settings:

- next to the Control Name, have a button to suggest a name from the label if the control name looks auto-generate
- in the Label, have a button to suggest a label from the name if the control name doesn't look auto-generate

Separately we could have a dialog to show all control names/labels to set names/labels quickly. There might be an RFE for something like this already.
